sdk/core/Azure.Core/src/Shared/ConnectionString.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core
+{
+    internal sealed class ConnectionString
+    {
+        private readonly Dictionary<string, string> _pairs;
+        private readonly string _pairSeparator;
+        private readonly string _keyValueSeparator;
+
+        public static ConnectionString Parse(string connectionString, string segmentSeparator = ";", string keyValueSeparator = "=", bool allowEmptyValues = false)
+        {
+            Validate(connectionString, segmentSeparator, keyValueSeparator, allowEmptyValues);
+            return new ConnectionString(ParseSegments(connectionString, segmentSeparator, keyValueSeparator), segmentSeparator, keyValueSeparator);
+        }
+
+        private ConnectionString(Dictionary<string, string> pairs, string pairSeparator, string keyValueSeparator)
+        {
+            _pairs = pairs;
+            _pairSeparator = pairSeparator;
+            _keyValueSeparator = keyValueSeparator;
+        }
+
+        public string GetRequired(string key) =>
+            _pairs.TryGetValue(key, out var value) ? value : throw new InvalidOperationException($"Required key '{key}' is missing in connection string.");
+
+        public string GetNonRequired(string key) =>
+            _pairs.TryGetValue(key, out var value) ? value : null;
+
+        public void Replace(string key, string value)
+        {
+            if (_pairs.ContainsKey(key))
+            {
+                _pairs[key] = value;
+            }
+        }
+
+        public override string ToString()
+        {
+            var stringBuilder = new StringBuilder();
+            var isFirst = true;
+            foreach (KeyValuePair<string, string> segment in _pairs)
+            {
+                if (isFirst)
+                {
+                    isFirst = false;
+                }
+                else
+                {
+                    stringBuilder.Append(_pairSeparator);
+                }
+
+                stringBuilder.Append(segment.Key);
+                if (segment.Value != null)
+                {
+                    stringBuilder.Append(_keyValueSeparator).Append(segment.Value);
+                }
+            }
+
+            return stringBuilder.ToString();
+        }
+
+
```

Comment:
nit: looks like there's an extra carriage return here.  Are static analyzers running on this?

Reply:
Fixed